File: go/dsa/matrix/SortedMatrix.go
package matrix

import "fmt"

// Search busca por um valor x em uma matriz ordenada.
// A busca começa no canto superior direito da matriz e segue
// para a esquerda ou para baixo, dependendo da comparação com x.
func Search(matrix [][]int, n int, x int) {
	// Inicializa as variáveis i e j. i começa na primeira linha (0),
	// e j começa na última coluna (n - 1).
	i := 0
	j := n - 1

	// Continua a busca enquanto i estiver dentro dos limites da matriz
	// e j não ultrapassar o limite esquerdo.
	for i < n && j >= 0 {
		// Se o elemento atual é igual a x, imprime a posição e sai da função.
		if matrix[i][j] == x {
			fmt.Printf("x found at - %v,%v\n", i, j)
			return
		}

		// Se o elemento atual é maior que x, move para a esquerda.
		if matrix[i][j] > x {
			j--
		} else {
			// Se o elemento atual é menor que x, move para baixo.
			i++
		}
	}
	// Se o loop terminar sem encontrar x, imprime que o valor não foi encontrado.
	fmt.Println("Value not found!")
}

File: csharp/src/dsa/matrix/SortedMatrix.cs
namespace dsa.matrix;

public class SortedMatrix
{
    // static void Main()
    // {
    //     int[][] matrix = {
    //         new int[] { 10, 20, 30, 40 },
    //         new int[] { 15, 25, 35, 45 },
    //         new int[] { 27, 29, 37, 48 },
    //         new int[] { 32, 33, 39, 51 }
    //     };

    //     SortedMatrix sm = new();
    //     sm.Search(matrix, matrix.Length, 32);
    //     sm.Search(matrix, matrix.Length, 100);
    // }

    public void Search(int[][] matrix, int n, int x)
    {
        // Iniciamos no topo direito da matriz, a posição mais estratégica para a busca.
        int i = 0;
        int j = n - 1;

        // Vamos percorrer a matriz enquanto estivermos dentro dos limites dela.
        while (i < n && j >= 0)
        {
            // Se encontrarmos o número que estamos procurando, mostramos onde ele está.
            if (matrix[i][j] == x)
            {
                Console.WriteLine($"x found at - {i},{j}");
                return;
            }
            // Se o número na posição atual for maior que o procurado, movemos para a esquerda.
            if (matrix[i][j] > x)
                j--;
            // Se for menor, descemos para a próxima linha.
            else
                i++;
        }
        // Se terminarmos a busca sem encontrar o número, avisamos que ele não está na matriz.
        Console.WriteLine("Value not found");
    }
}
